src/cluster/OnLoop.h

```diff
@@ -0,0 +1,175 @@
+// See the file "COPYING" in the main distribution directory for copyright.
+
+#pragma once
+
+#include <atomic>
+#include <chrono>
+#include <list>
+#include <mutex>
+#include <thread>
+
+#include "zeek/Flare.h"
+#include "zeek/Reporter.h"
+#include "zeek/iosource/IOSource.h"
+#include "zeek/iosource/Manager.h"
+
+namespace zeek::detail {
+/**
+ * Template class allowing work items to be queued by threads and processed
+ * in Zeek's main thread.
+ *
+ * This is similar to MsgThread::SendOut(), but doesn't require usage of BasicThread
+ * or MsgThread instances. Some libraries spawn their own threads or invoke callbacks
+ * from arbitrary threads. OnLoopProcess::QueueForProcessing() can be used to transfer
+ * work from such callbacks onto Zeek's main thread.
+ *
+ * There's currently no explicit way to transfer a result back. If this is needed,
+ * have the queueing thread block on a semaphore or condition variable and update
+ * it from Process().
+ *
+ * Note that QueueForProcessing() puts the queueing thread to sleep if there's
+ * too many items in the queue.
+ */
+template<class Proc, class Work>
+class OnLoopProcess : public zeek::iosource::IOSource {
```

Comment:
If I understand it correctly this is basically a message queue from threads into Zeek. I really like that abstraction, but what's the intuition for "OnLoop"? Is it that this feeds into Zeek's I/O loop, from some externally established thread, or that it runs in the I/O loop of such threads?

Reply:
> Is it that this feeds into Zeek's I/O loop, from some externally established thread,

This. And particularly running code on the main thread.

The name is not great, but it stuck.

I also had chatted with @bbannier and he was a bit afraid that everyone builds their own things to this. I acknowledge this. We settled somewhat on keeping an eye out if this happens elsewhere and unify or come up with something better. E.g. @Neverlord's recent logging changes have a similar class, to run logging in the main-thread, but the tasks come from within broker/caf threads. That's why it's in `src/cluster/` instead of somewhere more common.